---
title: "Incorporating Homer into LGCP ChIPSeq Pipeline"
author: "Brian Capaldo"
date: "3/14/2019"
output: github_document
---

# Homer

Homer appears to be one of the most recommended tools for peak annotation and motif finding. The intention is to update the current LGCP ChIPSeq nextflow based pipeline to incorporate homer. There are two functions that will be incorporated, `findMotifsGenome.pl`, and `annotatePeaks.pl`. The `annotatePeaks.pl` is the simpler of the two, and will cover that one first. Neither of the functions depend on the other, so rather than adding one process to handle both, we'll be adding two additional processes into [main.nf](../chipseq/main.nf), and cluster options for both of those new processes in the [biowulf config file](../chipseq/conf/biowulf-modules.config).  

From the docs:
> Region Size ("-size <#>", "-size <#>,<#>", "-size given", default: 200)

> The size of the region used for motif finding is important.  If analyzing ChIP-Seq peaks from a transcription factor, Chuck would recommend 50 bp for establishing the primary motif bound by a given transcription factor and 200 bp for finding both primary and "co-enriched" motifs for a transcription factor.  When looking at histone marked regions, 500-1000 bp is probably a good idea (i.e. H3K4me or H3/H4 acetylated regions).  In theory, HOMER can work with very large regions (i.e. 10kb), but with the larger the regions comes more sequence and longer execution time.  These regions will be based off the center of the peaks.  If you prefer an offset, you can specify "-size -300,100" to search a region of size 400 that is centered 100 bp upstream of the peak center (useful if doing motif finding on putative TSS regions).  If you have variable length regions, use the option "-size given" and HOMER will use the exact regions that were used as input.


```{java eval = F}
/*
 * STEP 10.2 Motif Finding with homer
 */

process homer_find_motifs {
   tag "${homer_bed.baseName}"
   publishDir "${params.outdir}/macs/homer", mode: 'copy',
       saveAs: {filename ->
            "${homer_bed.baseName}/" 
        }
 
   input:
   file homer_bed from macs_narrow_peak_homer
   
   output:
   file '*.{motifs,all.motifs,motifFindingParameters.txt,knownResults.txt,seq.autonorm.tsv,html}' into homer_motifs_results
   stdout channel
 
   script:
   """
   findMotifsGenome.pl $homer_bed hg19 ./ -size 200 -mask -preparse -preparsedDir ./ -p 4
   """
}
```

```{bash, eval=F}

#input_file=/data/khanlab/projects/ChIP_seq/data_by_file_type/test/DATA/Sample_RH4_H3K27ac_001_C_H5TLGBGXX/ROSE_out_narrow_1000/trimmed_AllEnhancers.sorted.table.txt_fpkm3.regular.bed
#output_dir=/data/khanlab/projects/ChIP_seq/data_by_file_type/test/DATA/Sample_RH4_H3K27ac_001_C_H5TLGBGXX/ROSE_out_narrow_1000/motif_regular
#preparse_dir=/data/khanlab/projects/ChIP_seq/data_by_file_type/test/scripts/preparse


findMotifsGenome.pl <peak/BED file> <genome> <output directory> -size 200 -mask -p <# threads> -preparse
  findMotifsGenome.pl $input_file $genome $output_dir -size $size -len $len -preparsedDir $preparsedDir -p $p

# Region Size ("-size <#>", "-size <#>,<#>", "-size given", default: 200)
# The size of the region used for motif finding is important.  If analyzing ChIP-Seq peaks from a transcription factor, Chuck would recommend 50 bp for establishing the primary motif bound by a given transcription factor and 200 bp for finding both primary and "co-enriched" motifs for a transcription factor.  When looking at histone marked regions, 500-1000 bp is probably a good idea (i.e. H3K4me or H3/H4 acetylated regions).  In theory, HOMER can work with very large regions (i.e. 10kb), but with the larger the regions comes more sequence and longer execution time.  These regions will be based off the center of the peaks.  If you prefer an offset, you can specify "-size -300,100" to search a region of size 400 that is centered 100 bp upstream of the peak center (useful if doing motif finding on putative TSS regions).  If you have variable length regions, use the option "-size given" and HOMER will use the exact regions that were used as input.

```

